param/parameterized.py

```diff
@@ -1318,29 +1317,29 @@ def _set_instantiate(self,instantiate):
             self.instantiate = self._slot_defaults['instantiate']
 
     def __setattr__(self, attribute, value):
-        if attribute == 'name' and getattr(self, 'name', None) and value != self.name:
-            raise AttributeError("Parameter name cannot be modified after "
-                                 "it has been bound to a Parameterized.")
-
-        implemented = (attribute != "default" and attribute in getattr(self, 'watchers', []))
-        slot_attribute = attribute in self.__class__._all_slots_
-        try:
-            old = getattr(self, attribute) if implemented else NotImplemented
-            if slot_attribute:
+        if attribute == 'name':
+            name = getattr(self, 'name', None)
+            if name is not None and value != name:
+                raise AttributeError("Parameter name cannot be modified after "
+                                     "it has been bound to a Parameterized.")
+
+        is_slot = attribute in self.__class__._all_slots_
+        has_watcher = attribute != "default" and attribute in getattr(self, 'watchers', [])
+        if not (is_slot or has_watcher):
```

Comment:
We could use the walrus operator here so we exit earlier and do not need the calculation of `has_watcher`. 

This is only possible if min. version is Python 3.8.

Reply:
py38 being four years old at this point, I'm happy with that as a version floor for Param 2.0, if that's ok for Panel purposes.

Reply:
@Hoxbro how would you use the walrus operator here?

I believe this could be rewritten as follows. But  I'm happy with how it is right now and haven't run any benchmark to see whether it'd make any difference.

```python
if not is_slot:
    has_watcher = attribute != "default" and attribute in getattr(self, 'watchers', [])
    if not has_watcher:
        # Return early if attribute is not a slot
        return super().__setattr__(attribute, value)
```

Oh and the minimum Python version for Param is 3.8 already :)

Reply:
I didn't put that many thoughts into it. 

Looking at this again. I don't think it is possible to do as both variables are needed if we don't exit early.

Your example will give an `UnboundLocalError` and similar with the Walrus operator code change.

Reply:
Ah indeed `has_watcher` is re-used below, my bad!